diff --git a/src/py2cpp/util/2to3/epydoc/cli.py b/src/py2cpp/util/2to3/epydoc/cli.py
index d0294ed..5cf6eb4 100644
--- a/src/py2cpp/util/2to3/epydoc/cli.py
+++ b/src/py2cpp/util/2to3/epydoc/cli.py
@@ -61,6 +61,9 @@ levels are currently defined as follows::
   1               bar             yes            yes       yes
   2               list            yes            yes       yes
 """
+
+from __future__ import print_function
+
 __docformat__ = 'epytext en'
 
 import sys, os, time, re, pickle, textwrap
diff --git a/src/py2cpp/util/2to3/epydoc/docwriter/html.py b/src/py2cpp/util/2to3/epydoc/docwriter/html.py
index 40d25c6..efb419c 100644
--- a/src/py2cpp/util/2to3/epydoc/docwriter/html.py
+++ b/src/py2cpp/util/2to3/epydoc/docwriter/html.py
@@ -12,6 +12,9 @@ this module is the L{HTMLWriter} class.
 
 @todo: Add a cache to L{HTMLWriter.url()}?
 """
+
+from __future__ import print_function
+
 __docformat__ = 'epytext en'
 
 import re, os, sys, codecs, sre_constants, pprint, base64
diff --git a/src/py2cpp/util/2to3/epydoc/docwriter/xlink.py b/src/py2cpp/util/2to3/epydoc/docwriter/xlink.py
index 49b33f6..9ab7480 100644
--- a/src/py2cpp/util/2to3/epydoc/docwriter/xlink.py
+++ b/src/py2cpp/util/2to3/epydoc/docwriter/xlink.py
@@ -67,6 +67,8 @@ API Linking Options::
 .. _Docutils: http://docutils.sourceforge.net/
 """
 
+from __future__ import print_function
+
 # $Id: xlink.py 1586 2007-03-14 01:53:42Z dvarrazzo $
 __version__ = "$Revision: 1586 $"[11:-2]
 __author__ = "Daniele Varrazzo"
diff --git a/src/py2cpp/util/2to3/epydoc/gui.py b/src/py2cpp/util/2to3/epydoc/gui.py
index 460d245..f948b45 100644
--- a/src/py2cpp/util/2to3/epydoc/gui.py
+++ b/src/py2cpp/util/2to3/epydoc/gui.py
@@ -28,6 +28,9 @@ Usage::
 @todo: Use ini-style project files, rather than pickles (using the
 same format as the CLI).
 """
+
+from __future__ import print_function
+
 __docformat__ = 'epytext en'
 
 import sys, os.path, re, glob
diff --git a/src/py2cpp/util/2to3/epydoc/markup/epytext.py b/src/py2cpp/util/2to3/epydoc/markup/epytext.py
index e70a020..a33b682 100644
--- a/src/py2cpp/util/2to3/epydoc/markup/epytext.py
+++ b/src/py2cpp/util/2to3/epydoc/markup/epytext.py
@@ -97,6 +97,8 @@ Description::
 # Note: the symbol list is appended to the docstring automatically,
 # below.
 
+from __future__ import print_function
+
 __docformat__ = 'epytext en'
 
 # Code organization..
